### PR TITLE
[SYCL] has_extension free functions

### PR DESCRIPTION
**Pull request by andylshort**
Implementation of the `has_extension` free functions defined in the SYCL 2020 spec for the `sycl::opencl` namespace:  https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_extension_query

As `info::device::extensions` is being deprecated in favour of `aspects`, it was decided to implement these with direct calls to the Plugin API instead.

### CODE REVIEW

File: sycl/source/backend/opencl.cpp
Comment (romanovvlad) [inline sycl/source/backend/opencl.cpp:81]: Can't we just search Extension in the original `std::string` like this?:
```suggestion
  return ExtensionsList.find(Extension) != std::string::npos;
```

**Review by romanovvlad — CHANGES_REQUESTED** [review]

File: sycl/source/backend/opencl.cpp
Comment (aelovikov-intel) [reply sycl/source/backend/opencl.cpp:71]: I don't think so, but we can use `std::string_view ExtensionsString` to avoid copy and still get the `operator==` for use in `std::find`.

Comment (romanovvlad) [reply sycl/source/backend/opencl.cpp:71]: >> I don't think so,

Could you please clarify why?

>> but we can use std::string_view ExtensionsString to avoid

Isn't `std::string_view` is a non-owning container?

Comment (aelovikov-intel) [reply sycl/source/backend/opencl.cpp:71]: Even if std::string would work, that would be very non-conventional use of filling it by a C-style API by providing that API with a raw pointer into it which I don't consider a good solution. I'd very much prefer the current approach of having a smart pointer for the memory allocation instead.

> Isn't std::string_view a non-owning container? 

Yes, that's why we'll still need this `Result` smart-pointer to own the memory. But we'd be able to use the view for subsequent manipulations without creating copies.

Comment (andylshort) [reply sycl/source/backend/opencl.cpp:71]: I used the logic from the existing platform/device info retrieval methods as I thought it would be better to keep it as self-contained as possible and thought that process was standard procedure. 

- Device: https://github.com/intel/llvm/blob/sycl/sycl/source/detail/device_info.hpp#L160
- Platform: https://github.com/intel/llvm/blob/sycl/sycl/source/detail/platform_info.hpp#L21

What are your recommendations? Keep as-is, use the above inline helpers, or rewrite the existing with string_view?

Comment (romanovvlad) [reply sycl/source/backend/opencl.cpp:71]: `string_view`

File: sycl/source/backend/opencl.cpp
Comment (AlexeySachkov) [inline sycl/source/backend/opencl.cpp:51]: Could you please clarify what this `TODO` means, i.e. what is the action we supposed to take in the future? My understanding is that even though device extensions are deprecated in SYCL in favor of device aspects, OpenCL extensions are not going anywhere and for that we have this backend-specific interface.

File: sycl/source/backend/opencl.cpp
Comment (AlexeySachkov) [inline sycl/source/backend/opencl.cpp:71]: Technically, this goes against our coding standards: [Don’t Use Braces on Simple Single-Statement Bodies of if/else/loop Statements](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements)

File: sycl/test/abi/sycl_symbols_linux.dump
Comment (AlexeySachkov) [inline sycl/test/abi/sycl_symbols_linux.dump:3899]: I suppose [`sycl_symbols_windows.dump`](https://github.com/intel/llvm/blob/sycl/sycl/test/abi/sycl_symbols_windows.dump) should be updated as well? It is just we don't have Windows build and tests in our pre-commit and therefore it would only be visible in post-commit that some tests fail

File: sycl/source/backend/opencl.cpp
Comment (AlexeySachkov) [inline sycl/source/backend/opencl.cpp:89]: FYI: [LLVM coding standard](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable) says:

> Use `auto` if and only if it makes the code more readable or easier to maintain. Don’t “almost always” use `auto`...

I wouldn't say that type of `DeviceImpl` can be easily inferred just by looking at this line of code, but I don't have strong opinion here (and I'm not a codeowner here), so I will leave it up to others whether or not it is ok to use `auto` here.

File: sycl/source/backend/opencl.cpp
Comment (aelovikov-intel) [inline sycl/source/backend/opencl.cpp:94]:
```suggestion
  // info::device::extensions call.
```

File: sycl/source/backend/opencl.cpp
Comment (steffenlarsen) [inline sycl/source/backend/opencl.cpp:62]:
```suggestion
  plugin &Plugin = PlatformImpl->getPlugin();
```

File: sycl/source/backend/opencl.cpp
Comment (dm-vodopyanov) [inline sycl/source/backend/opencl.cpp:68]: nit: 
```suggestion
      PluginPlatform, PI_PLATFORM_INFO_EXTENSIONS, /*param_value_size=*/0, /*param_value=*/nullptr, &ResultSize);
```

File: sycl/source/backend/opencl.cpp
Comment (andylshort) [reply sycl/source/backend/opencl.cpp]: I will review this document and make sure my code better adheres to the standards.

File: sycl/source/backend/opencl.cpp
Comment (andylshort) [reply sycl/source/backend/opencl.cpp]: Changed to the full type of `std::shared_ptr<sycl::detail::device_impl>` (and platform equiv.) for clarity.

File: sycl/source/backend/opencl.cpp
Comment (andylshort) [reply sycl/source/backend/opencl.cpp]: This required adding a `const` qualifier, so did this change and that in both methods.

File: sycl/include/sycl/backend/opencl.hpp
Comment (andylshort) [reply sycl/include/sycl/backend/opencl.hpp:67]: Rebased to fix this, hence the spam below. Resolved now.

File: sycl/include/sycl/backend/opencl.hpp
Comment (premanandrao) [inline sycl/include/sycl/backend/opencl.hpp:70]: Nit: looks like missing newline at EOF.

Comment (andylshort) [reply sycl/include/sycl/backend/opencl.hpp]: Resolved with rebase to add back.

File: sycl/include/sycl/backend/opencl.hpp
Comment (andylshort) [reply sycl/include/sycl/backend/opencl.hpp]: Removed extra newline with rebase

File: sycl/include/sycl/backend/opencl.hpp
Comment (romanovvlad) [inline sycl/include/sycl/backend/opencl.hpp:33]:
```suggestion
__SYCL_EXPORT bool has_extension(const sycl::platform &SyclPlatform,
                                 const std::string &Extension);
__SYCL_EXPORT bool has_extension(const sycl::device &SyclDevice,
                                 const std::string &Extension);
```
Minor.

**Review by romanovvlad — APPROVED** [review]

File: sycl/include/sycl/backend/opencl.hpp
Comment (andylshort) [reply sycl/include/sycl/backend/opencl.hpp]: I have now resolved this in commit [b2e2c43](https://github.com/intel/llvm/pull/7059/commits/b2e2c43210194e3534ca24121395c4d6adf2efe3) that I have just rebased.

**Review by romanovvlad — APPROVED** [review]

File: sycl/unittests/SYCL2020/HasExtension.cpp
Comment (andylshort) [reply sycl/unittests/SYCL2020/HasExtension.cpp:28]: Oh, and EXPECT_FALSE that extension? I can do that, if so.

**Review by romanovvlad — APPROVED** [review]

**Review by steffenlarsen — APPROVED** [review]

**Pull request merged**
